crates/concrete_check/src/linearity_check.rs

```diff
@@ -0,0 +1,1086 @@
+                var
+            );
+        }
+        self.vars.get_mut(var)
+    }
+
+    fn get_info(&self, var: &str) -> Option<&VarInfo> {
+        self.vars.get(var)
+    }
+
+    // Retrieve a variable's state
+    fn _get_state(&mut self, var: &str) -> Option<&VarState> {
+        if let Some(info) = self.get_info(var) {
+            Some(&info.state)
+        } else {
+            None
+        }
+    }
+
+    // Retrieve a variable's state
+    fn update_state(&mut self, var: &str, new_state: &VarState) {
+        let info = self.get_info_mut(var);
+        if let Some(info) = info {
+            //info.state = new_state.clone();
+            info.state = *new_state;
+        }
+    }
+
+    fn get_loop_depth(&mut self, name: &str) -> usize {
+        let state = self.get_info(name);
+        if let Some(state) = state {
+            state.depth
+        } else {
+            0
+        }
+    }
+
+    pub fn _tables_are_consistent(&self, other: &StateTbl) -> Result<(), LinearityError> {
+        for (key, value) in &self.vars {
+            if let Some(other_value) = other.vars.get(key) {
+                if value != other_value {
+                    return Err(LinearityError::StateInconsistency {
+                        message: format!(
+                            "Variable '{}' state mismatch: {:?} != {:?}",
+                            key, value, other_value
+                        ),
+                    });
+                }
+            } else {
+                return Err(LinearityError::StateInconsistency {
+                    message: format!("Variable '{}' not found in the other table", key),
+                });
+            }
+        }
+        // Also check for variables in `other` not present in `self`
+        for key in other.vars.keys() {
+            if !self.vars.contains_key(key) {
+                return Err(LinearityError::StateInconsistency {
+                    message: format!("Variable '{}' not found in the first table", key),
+                });
+            }
+        }
+
+        Ok(())
+    }
+}
+
+struct LinearityChecker {
+    //state_tbl: StateTbl,
+}
+
+#[allow(dead_code)]
+#[allow(unused_variables)]
+impl LinearityChecker {
+    fn new() -> Self {
+        LinearityChecker {
+            //state_tbl: StateTbl::new(),
+        }
+    }
+
+    fn consume_once(
+        &self,
+        mut state_tbl: StateTbl,
+        depth: usize,
+        name: &str,
+    ) -> Result<StateTbl, LinearityError> {
+        let loop_depth = state_tbl.get_loop_depth(name);
+        tracing::debug!(
+            "Consuming variable: {} depth {} loop_depth {}",
+            name,
+            depth,
+            loop_depth
+        );
+        if depth == state_tbl.get_loop_depth(name) {
+            state_tbl.update_state(name, &VarState::Consumed);
+            tracing::debug!("Consumed variable: {}", name);
+            /*
+            let mut state = self.state_tbl.get_state(name);
```

Comment:
can you add a short comment explaining why these lines are commented?

Reply:
Removed the commented code as it was an approach not implemented. Implemented specific error when a variable created outside a loop is consumed inside the loop.